onnxruntime/core/session/inference_session.h

```diff
@@ -212,85 +128,212 @@ class InferenceSession {
     * to get the name/type/shapes of the inputs.
     * @return pair.first = OK; FAIL otherwise. pair.second is non-NULL when pair.first = OK.
     * @note lifetime of the returned pointer is valid as long as the Session object is live.
     */
-  std::pair<common::Status, const InputDefList*> GetModelInputs() const;
+  std::pair<common::Status, const InputDefList*> GetModelInputs() const override;
 
   /**
     * Get all output definitions of the model. Use this to get the name/type/shapes of the outputs.
     * @return pair.first = OK; FAIL otherwise. pair.second is non-NULL when pair.first = OK.
     * @note lifetime of the returned pointer is valid as long as the Session object is live.
     */
-  std::pair<common::Status, const OutputDefList*> GetModelOutputs() const;
+  std::pair<common::Status, const OutputDefList*> GetModelOutputs() const override;
 
   /**
     * Get the current number of in-progress concurrent Run calls.
     */
-  int GetCurrentNumRuns();
+  int GetCurrentNumRuns() const override {
+    return current_num_runs_.load();
+  }
 
   /**
     * Start profiling on this inference session. This simply turns on profiling events to be 
     * recorded. A corresponding EndProfiling has to follow to write profiling data to a file.
     *@param file_prefix is the prefix of the profile file. It can include a directory path. 
     */
-  void StartProfiling(const std::string& file_prefix);
+  void StartProfiling(const std::string& file_prefix) override;
 #ifdef _WIN32
-  void StartProfiling(const std::wstring& file_prefix);
+  void StartProfiling(const std::wstring& file_prefix) override;
 #endif
   /**
     * Start profiling on this inference session. This simply turns on profiling events to be
     * recorded. A corresponding EndProfiling has to follow to send profiling events through the logger's ISink.
     *@param logger_ptr is pointer to the logger where profiling events will be sent to.
     */
-  void StartProfiling(const logging::Logger* logger_ptr);
+  void StartProfiling(const logging::Logger* logger_ptr) override;
 
   /**
     * Write captured profile events in chromium format.
     @return the name of the profile file.
     */
-  std::string EndProfiling();
+  std::string EndProfiling() override;
 
  protected:
   /**
     * Load an ONNX model.
     * @param protobuf object corresponding to the model file. model_proto will be copied by the API.
     * @return OK if success.
     */
-  common::Status Load(const ONNX_NAMESPACE::ModelProto& model_proto);
+  common::Status Load(const ONNX_NAMESPACE::ModelProto& model_proto) override;
 
   /**
     * Load an ONNX model.
     * @param protobuf object corresponding to the model file. This is primarily to support large models.
     * @return OK if success.
     */
-  common::Status Load(std::unique_ptr<ONNX_NAMESPACE::ModelProto> p_model_proto);
+  common::Status Load(std::unique_ptr<ONNX_NAMESPACE::ModelProto> p_model_proto) override;
 
  private:
+  friend class Session;
```

Comment:
>friend class Session; [](start = 2, length = 21)

nit: Friendship is needed for instantiation? I'd get rid of it as follows:
Declare ctor to take in a ref or a ptr of a forward declared structure that is only available supplied by Session class. This way even a public constructor can be called by anyone else.

Reply:
This is exquisite. I'll try it in next iteration, thanks.